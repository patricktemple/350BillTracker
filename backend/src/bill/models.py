import enum
from uuid import uuid4

import flask
from sqlalchemy import Boolean, Column, Enum, ForeignKey, Integer, Text, sql
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.ext.declarative import declared_attr
from sqlalchemy.orm import foreign, relationship, remote

from ..models import TIMESTAMP, UUID, db
from ..user.models import User
from ..utils import now

DEFAULT_TWITTER_SEARCH_TERMS = [
    "solar",
    "climate",
    "wind power",
    "renewable",
    "fossil fuel",
]


class StateChamber(enum.Enum):
    SENATE = 1
    ASSEMBLY = 2


class Bill(db.Model):
    """
    Base table for all bills, both city and state. Contains any info that's
    relevant for both kinds of bills.

    Bills are polymorphic. This table has a "type" which specifies whether it's
    a city or a state bill, and there will be an associated row in the CityBill
    and StateBill with more info on it."""

    __tablename__ = "bills"

    class BillType(enum.Enum):
        CITY = 1
        STATE = 2

    id = Column(UUID, primary_key=True, default=uuid4)

    # For city this is the "MatterName", for state it is the "title"
    name = Column(Text, nullable=False)

    # For city this is the "title", for state this is the "summary"
    description = Column(Text, nullable=False)

    # Info on child objects:
    type = Column(Enum(BillType), nullable=False)

    # If this is is a city bill, points to more specific data on that.
    city_bill = relationship(
        "CityBill",
        back_populates="bill",
        uselist=False,
        cascade="all, delete",
        lazy="joined",
    )

    # If this is is a state bill, points to more specific data on that.
    state_bill = relationship(
        "StateBill",
        back_populates="bill",
        uselist=False,
        cascade="all, delete",
        lazy="joined",
    )

    # Data we track
    notes = Column(Text, nullable=False, server_default="")
    nickname = Column(Text, nullable=False, server_default="")
    twitter_search_terms = Column(
        ARRAY(Text), nullable=False, default=DEFAULT_TWITTER_SEARCH_TERMS
    )

    power_hours = relationship(
        "PowerHour", back_populates="bill", cascade="all, delete"
    )
    attachments = relationship(
        "BillAttachment", back_populates="bill", cascade="all, delete"
    )

    @property
    def display_name(self):
        return self.nickname if self.nickname else self.name

    @property
    def tracked(self):
        return True

    @property
    def status(self):
        if self.type == Bill.BillType.CITY:
            return self.city_bill.status

        senate_status = (
            self.state_bill.senate_bill.status
            if self.state_bill.senate_bill
            else "(No Senate bill)"
        )
        assembly_status = (
            self.state_bill.assembly_bill.status
            if self.state_bill.assembly_bill
            else "(No Assembly bill)"
        )

        return f"{senate_status} / {assembly_status}"

    @property
    def code_name(self):
        if self.type == Bill.BillType.CITY:
            return self.city_bill.file

        senate_print_no = (
            self.state_bill.senate_bill.base_print_no
            if self.state_bill.senate_bill
            else "(No Senate bill)"
        )
        assembly_print_no = (
            self.state_bill.assembly_bill.base_print_no
            if self.state_bill.assembly_bill
            else "(No Assembly bill)"
        )

        return f"{senate_print_no} / {assembly_print_no} from {self.state_bill.session_year} session"
    
    user_bill_settings = relationship("UserBillSettings", back_populates="bill", cascade="all, delete-orphan")


class BillAttachment(db.Model):
    __tablename__ = "bill_attachments"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(UUID, ForeignKey("bills.id"), nullable=False, index=True)
    bill = relationship("Bill", back_populates="attachments")

    name = Column(Text)
    url = Column(Text, nullable=False)


class PowerHour(db.Model):
    __tablename__ = "power_hours"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(UUID, ForeignKey("bills.id"), nullable=False, index=True)

    title = Column(Text)
    spreadsheet_url = Column(Text, nullable=False)
    spreadsheet_id = Column(Text, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)
    bill = relationship("Bill", back_populates="power_hours")


class CityBill(db.Model):
    """
    City-specific details about an NYC bill. There must also be an associated
    row in the Bill parent table that has more general bill info (like name).
    """

    __tablename__ = "city_bills"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)

    # ID in the City Council API
    city_bill_id = Column(Integer, nullable=False, unique=True, index=True)

    # Code name often used to discuss the bill, like Int 2317-2021
    file = Column(Text, nullable=False)

    # The parent Bill object that represents this.
    bill = relationship("Bill", back_populates="city_bill", lazy="joined")

    intro_date = Column(TIMESTAMP, nullable=False)

    status = Column(Text, nullable=False)

    # Which amendment version is currently active. We only track one at a time,
    # and sponsorships are tied to that particular version.
    active_version = Column(Text, nullable=False)

    sponsorships = relationship(
        "CitySponsorship", back_populates="city_bill", cascade="all, delete"
    )

    # Committee name
    council_body = Column(Text)


# TODO: Understand bill substitution and change if needed
class StateBill(db.Model):
    """
    State-specific details about a state bill. There must also be an associated
    row in the Bill parent table that has more general bill info (like name)."""

    __tablename__ = "state_bills"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)
    bill = relationship(Bill, back_populates="state_bill", lazy="joined")

    # The start of the 2-year legislative session this belongs to.
    session_year = Column(Integer, nullable=False)

    senate_bill = relationship(
        "SenateBill",
        back_populates="state_bill",
        uselist=False,
        cascade="all, delete",
    )
    assembly_bill = relationship(
        "AssemblyBill",
        back_populates="state_bill",
        uselist=False,
        cascade="all, delete",
    )


class StateChamberMixin:
    """
    Mixin for a bill in either the Senate or Assembly chamber. The have
    identical structure but are kept in separate tables."""

    active_version = Column(Text, nullable=False)
    status = Column(Text, nullable=False)
    base_print_no = Column(Text, nullable=False)

    @declared_attr
    def bill_id(self):
        return Column(UUID, ForeignKey(StateBill.bill_id), primary_key=True)

    # The Senate and the Assembly each run separate websites, and each website can
    # lookup both senate and assembly bills. So they're redundant websites with
    # very different UI. Therefore senate and assembly bills each have a
    # both assembly and senate websites.
    @property
    def senate_website(self):
        return f"https://www.nysenate.gov/legislation/bills/{self.state_bill.session_year}/{self.base_print_no}"

    @property
    def assembly_website(self):
        return f"https://nyassembly.gov/leg/?term={self.state_bill.session_year}&bn={self.base_print_no}"


class SenateBill(db.Model, StateChamberMixin):
    __tablename__ = "senate_bills"

    state_bill = relationship(StateBill, back_populates="senate_bill")
    sponsorships = relationship(
        "SenateSponsorship",
        back_populates="chamber_bill",
        cascade="all, delete-orphan",
    )


class AssemblyBill(db.Model, StateChamberMixin):
    __tablename__ = "assembly_bills"

    state_bill = relationship(StateBill, back_populates="assembly_bill")
    sponsorships = relationship(
        "AssemblySponsorship",
        back_populates="chamber_bill",
        cascade="all, delete-orphan",
    )


class UserBillSettings(db.Model):
    __tablename__ = "user_bill_settings"

    bill_id = Column(UUID, ForeignKey(Bill.id), primary_key=True)
    user_id = Column(UUID, ForeignKey(User.id), primary_key=True)

    send_bill_update_notifications = Column(
        Boolean, nullable=False, server_default=sql.false()
    )

    user = relationship(User, lazy="joined", back_populates="bill_settings")
    bill = relationship(Bill, back_populates="user_bill_settings")


Bill.viewer_settings = relationship(
    UserBillSettings,
    primaryjoin=(remote(UserBillSettings.bill_id) == foreign(Bill.id))
    & (
        UserBillSettings.user_id
        == sql.bindparam(
            "request_user_id", callable_=lambda: flask.g.request_user_id
        )
    ),
    viewonly=True,
)
