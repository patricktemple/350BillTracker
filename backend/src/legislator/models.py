from uuid import uuid4

from sqlalchemy import Column, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from ..models import TIMESTAMP, UUID, db


class Legislator(db.Model):
    __tablename__ = "legislators"

    # These come from the API
    id = Column(Integer, primary_key=True)
    name = Column(Text, nullable=False)
    term_start = Column(TIMESTAMP)
    term_end = Column(TIMESTAMP)
    email = Column(Text)
    district_phone = Column(Text)
    legislative_phone = Column(Text)

    sponsorships = relationship("BillSponsorship", back_populates="legislator")
    staffers = relationship("Staffer", back_populates="legislator")

    borough = Column(Text)
    website = Column(Text)

    # These are added by our static data
    twitter = Column(Text)  # exclude the @ symbol
    party = Column(Text)

    # Track our own info on the bill.
    notes = Column(Text)

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def twitter_url(self):
        return (
            f"https://www.twitter.com/{self.twitter}" if self.twitter else None
        )


class Staffer(db.Model):
    __tablename__ = "staffers"

    id = Column(UUID, primary_key=True, default=uuid4)
    name = Column(Text, nullable=False)
    title = Column(Text)
    email = Column(Text)
    phone = Column(Text)
    legislator_id = Column(
        Integer, ForeignKey("legislators.id"), nullable=False
    )
    twitter = Column(Text)

    legislator = relationship("Legislator", back_populates="staffers")

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def display_string(self):
        contact_methods = [self.phone, self.email, self.display_twitter]
        contact_methods = [c for c in contact_methods if c]
        contact_string = ", ".join(contact_methods)
        if not contact_string:
            contact_string = "No contact info"
        title_string = f"{self.title} - " if self.title else ""

        return f"{title_string}{self.name} ({contact_string})"
