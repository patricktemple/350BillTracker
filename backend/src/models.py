from uuid import uuid4

from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    ForeignKey,
    Integer,
    Text,
    TypeDecorator,
    sql,
)
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.dialects.postgresql import TIMESTAMP as _TIMESTAMP
from sqlalchemy.dialects.postgresql import UUID as _UUID
from sqlalchemy.orm import relationship

from .app import app
from .utils import now

db = SQLAlchemy(app)


class UUID(TypeDecorator):
    impl = _UUID(as_uuid=True)


class TIMESTAMP(TypeDecorator):
    impl = _TIMESTAMP(timezone=True)


# TODO: UUIDs for some PKs?
class BillAttachment(db.Model):
    __tablename__ = "bill_attachments"

    id = Column(Integer, primary_key=True)
    bill_id = Column(
        Integer, ForeignKey("bills.id"), nullable=False, index=True
    )
    bill = relationship("Bill", back_populates="attachments")

    name = Column(Text)
    url = Column(Text, nullable=False)


class PowerHour(db.Model):
    __tablename__ = "power_hours"

    id = Column(UUID, primary_key=True, default=uuid4)
    bill_id = Column(
        Integer, ForeignKey("bills.id"), nullable=False, index=True
    )

    title = Column(Text)
    spreadsheet_url = Column(Text, nullable=False)
    spreadsheet_id = Column(Text, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)
    bill = relationship("Bill", back_populates="power_hours")
