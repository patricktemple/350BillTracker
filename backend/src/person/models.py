import enum
from uuid import uuid4

from sqlalchemy import Column, Enum, ForeignKey, Integer, Text
from sqlalchemy.orm import relationship

from ..models import TIMESTAMP, UUID, db


class Person(db.Model):
    """
    Base table for every kind of person we track. General info shared by all
    people, such as name and phone, are kept here.

    A person is polymorphic. There is also a "type" that specifies which kind
    of person this is (such as a senator), and there are
    separate tables for each type of person that has more specific details.
    """

    __tablename__ = "persons"

    class PersonType(enum.Enum):
        # City
        COUNCIL_MEMBER = 1

        # State
        ASSEMBLY_MEMBER = 2
        SENATOR = 3

        # General
        STAFFER = 4

    # This is the internal ID shared by all people
    id = Column(UUID, primary_key=True, default=uuid4)

    name = Column(Text, nullable=False)
    title = Column(Text)
    email = Column(Text)
    phone = Column(Text)

    twitter = Column(Text)  # excludes the @ symbol

    # Track our own info on the person
    notes = Column(Text)

    type = Column(Enum(PersonType), nullable=False)

    # If this person is a council member, points to the object that has more specific
    # details about them in that role.
    council_member = relationship(
        "CouncilMember", back_populates="person", uselist=False, lazy="joined"
    )

    # If this person is a state senator, points to the object that has more specific
    # details about them in that role.
    senator = relationship(
        "Senator", back_populates="person", uselist=False, lazy="joined"
    )

    # If this person is a state assembly, points to the object that has more specific
    # details about them in that role.
    assembly_member = relationship(
        "AssemblyMember", back_populates="person", uselist=False, lazy="joined"
    )

    office_contacts = relationship("OfficeContact", back_populates="person")

    # These are added by our static data
    # TODO: Make this an enum?
    party = Column(Text)

    @property
    def display_twitter(self):
        return "@" + self.twitter if self.twitter else None

    @property
    def twitter_url(self):
        return (
            f"https://www.twitter.com/{self.twitter}" if self.twitter else None
        )


class CouncilMember(db.Model):
    """
    Data about a specific council member. For each row in this there must also
    be a Person that has more general info about this person, such as name.
    """

    __tablename__ = "council_members"

    # Foreign key to Person parent table
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    # ID of the City Council API object, which is also called Person
    city_council_person_id = Column(
        Integer, nullable=False, index=True, unique=True
    )

    term_start = Column(TIMESTAMP)
    term_end = Column(TIMESTAMP)

    # District phone is stored in Person.phone
    legislative_phone = Column(Text)

    sponsorships = relationship(
        "CitySponsorship", back_populates="council_member"
    )

    # The parent Person object.
    person = relationship(
        "Person", back_populates="council_member", lazy="joined"
    )

    borough = Column(Text)
    website = Column(Text)


class Senator(db.Model):
    """
    Data about a specific state senator. For each row in this there must also
    be a Person that has more general info about this person, such as name.
    """

    __tablename__ = "senators"

    # Foreign key to Person parent table
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    # The parent Person object.
    person = relationship("Person", back_populates="senator", lazy="joined")

    sponsorships = relationship("SenateSponsorship", back_populates="senator")

    state_member_id = Column(Integer, nullable=False, unique=True)

    district = Column(Integer)

    @property
    def website(self):
        return (
            f"https://www.nysenate.gov/district/{self.district}"
            if self.district
            else None
        )


class AssemblyMember(db.Model):
    """
    Data about a specific state assembly member. For each row in this there must
    also be a Person that has more general info about this person, such as name.
    """

    __tablename__ = "assembly_members"

    # Foreign key to Person parent table
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    # The parent Person object.
    person = relationship(
        "Person", back_populates="assembly_member", lazy="joined"
    )

    state_member_id = Column(Integer, nullable=False, unique=True)

    sponsorships = relationship(
        "AssemblySponsorship", back_populates="assembly_member"
    )

    district = Column(Integer)

    @property
    def website(self):
        return (
            f"https://nyassembly.gov/mem/?ad={self.district}"
            if self.district
            else None
        )


# Staffers have a single boss. They're one to many.
class Staffer(db.Model):
    __tablename__ = "staffers"

    # ID of the person themselves:
    person_id = Column(UUID, ForeignKey(Person.id), primary_key=True)

    boss_id = Column(UUID, ForeignKey(Person.id), nullable=False, index=True)

    # The Person object that represents general info about this person (NOT the person that
    # they work for.)
    person = relationship(
        "Person",
        foreign_keys=[person_id],
        back_populates="staffer",
        lazy="joined",
    )

    # The Person that this staffer works for.
    boss = relationship("Person", foreign_keys=[boss_id])


# If this person is a Staffer, gets the Staffer object that represents them.
Person.staffer = relationship(
    Staffer,
    foreign_keys=[Staffer.person_id],
    back_populates="person",
    cascade="all, delete",
    uselist=False,
)

# Gets all the Persons that work for this person as staffers.
Person.staffer_persons = relationship(
    Person,
    secondary="staffers",
    primaryjoin=Person.id == Staffer.boss_id,
    secondaryjoin=Staffer.person_id == Person.id,
    viewonly=True,
)


class OfficeContact(db.Model):
    __tablename__ = "office_contacts"

    id = Column(UUID, primary_key=True, default=uuid4)

    person_id = Column(UUID, ForeignKey(Person.id), index=True)

    # address type enum, maybe?

    phone = Column(Text)
    fax = Column(Text)
    city = Column(Text)

    person = relationship(
        "Person", back_populates="office_contacts", lazy="joined"
    )