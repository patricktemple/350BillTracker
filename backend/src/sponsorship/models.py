from uuid import uuid4

from sqlalchemy import Column, ForeignKey, Integer, UniqueConstraint, select, func
from sqlalchemy.orm import foreign, relationship, remote, column_property

from ..bill.models import (
    AssemblyBill,
    Bill,
    CityBill,
    SenateBill,
    StateChamberMixin,
)
from ..models import TIMESTAMP, UUID, db
from ..person.models import AssemblyMember, CouncilMember, Person, Senator


class CitySponsorship(db.Model):
    __tablename__ = "city_sponsorships"

    id = Column(UUID, primary_key=True, default=uuid4)

    bill_id = Column(
        UUID,
        ForeignKey("city_bills.bill_id"),
        nullable=False,
        index=True,
    )
    city_bill = relationship(
        CityBill,
        back_populates="sponsorships",
    )

    council_member_id = Column(
        UUID,
        ForeignKey("council_members.person_id"),
        nullable=False,
        index=True,
    )
    council_member = relationship(
        CouncilMember,
        back_populates="sponsorships",
    )

    sponsor_sequence = Column(Integer, nullable=False)

    # The timestamp when we first saw this sponsorship in the bill's list.
    # This is a proxy for when the sponsor actually signed on to the bill.
    # Note that when we first start tracking a bill, it may already have sponsorships
    # and we don't know the date that those were added. We leave added_at as null,
    # in that case, and only fill this in for sponsorships that were added later on.
    added_at = Column(TIMESTAMP)

    __table_args__ = (
        UniqueConstraint(
            "bill_id", "council_member_id", name="_bill_council_member_unique"
        ),
    )


CitySponsorship.bill = relationship(
    Bill,
    primaryjoin=remote(Bill.id) == foreign(CitySponsorship.bill_id),
    viewonly=True,
)
CitySponsorship.person = relationship(
    Person,
    primaryjoin=remote(Person.id)
    == foreign(CitySponsorship.council_member_id),
    viewonly=True,
)


class SenateSponsorship(db.Model):
    __tablename__ = "senate_sponsorships"

    id = Column(UUID, primary_key=True, default=uuid4)

    senate_version_id = Column(
        UUID, ForeignKey(SenateBill.id), nullable=False, index=True
    )
    senate_version = relationship(
        SenateBill,
        back_populates="sponsorships",
    )

    senator_id = Column(
        UUID, ForeignKey(Senator.person_id), nullable=False, index=True
    )
    senator = relationship(Senator, back_populates="sponsorships")

    # TODO: Add sponsor_sequence like we have with city, if needed
    # TODO: Unique constraint


class AssemblySponsorship(db.Model):
    __tablename__ = "assembly_sponsorships"

    id = Column(UUID, primary_key=True, default=uuid4)

    assembly_version_id = Column(
        UUID, ForeignKey(AssemblyBill.id), nullable=False, index=True
    )
    assembly_version = relationship(
        AssemblyBill,
        back_populates="sponsorships",
    )

    assembly_member_id = Column(
        UUID, ForeignKey(AssemblyMember.person_id), nullable=False, index=True
    )
    assembly_member = relationship(
        AssemblyMember, back_populates="sponsorships"
    )

    # TODO: Add sponsor_sequence like we have with city, if needed
    # TODO: Unique constraint


# is this inefficient? it loads it fully for every single  element
CityBill.sponsor_count = column_property(select(func.count(CitySponsorship.id)).where(CitySponsorship.bill_id==CityBill.bill_id).scalar_subquery())
SenateBill.sponsor_count = column_property(select(func.count(SenateSponsorship.id)).where(SenateSponsorship.senate_version_id==SenateBill.id).scalar_subquery())
AssemblyBill.sponsor_count = column_property(select(func.count(AssemblySponsorship.id)).where(AssemblySponsorship.assembly_version_id==AssemblyBill.id).scalar_subquery())