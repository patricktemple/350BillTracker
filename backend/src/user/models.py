from uuid import uuid4

from sqlalchemy import Boolean, CheckConstraint, Column, ForeignKey, Text, sql
from sqlalchemy.orm import relationship

from ..models import TIMESTAMP, UUID, db
from ..utils import now


class User(db.Model):
    __tablename__ = "users"
    id = Column(UUID, primary_key=True, default=uuid4)

    email = Column(
        Text, index=True, nullable=False, unique=True
    )  # always lowercase
    name = Column(Text)

    login_links = relationship(
        "LoginLink", back_populates="user", cascade="all, delete-orphan"
    )

    # The "root" user can never be deleted.
    can_be_deleted = Column(Boolean, nullable=False, server_default=sql.true())

    bill_settings = relationship(
        "UserBillSettings", back_populates="user", cascade="all, delete-orphan"
    )

    bills_with_notifications = relationship(
        "Bill",
        secondary="user_bill_settings",
        primaryjoin="and_(UserBillSettings.user_id==User.id, UserBillSettings.send_bill_update_notifications)",
        secondaryjoin="UserBillSettings.bill_id==Bill.id",
        viewonly=True,
    )

    __table_args__ = (
        CheckConstraint(
            "email = lower(email)", name="check_email_is_lowercase"
        ),
    )


class LoginLink(db.Model):
    __tablename__ = "login_links"

    user_id = Column(UUID, ForeignKey(User.id), nullable=False)

    token = Column(Text, nullable=False, primary_key=True)

    expires_at = Column(TIMESTAMP, nullable=False)

    created_at = Column(TIMESTAMP, nullable=False, default=now)

    used_at = Column(TIMESTAMP)

    user = relationship("User", back_populates="login_links")

    # TODO: Consider only allowing these links to be used once. Better security
    # in case of leaked browser URL, but worse UX.


# TODO: Make these linkes only usably once
# We would need to have a good error page for this too, if they click it a second time
