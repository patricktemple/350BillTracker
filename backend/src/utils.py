import logging
from datetime import datetime, timezone

from src.models import db


def now():
    return datetime.utcnow().replace(tzinfo=timezone.utc)


def cron_function(func):
    def impl(*args, **kwargs):
        try:
            func(*args, **kwargs)
        except Exception:
            db.session.rollback()
            logging.exception("Exception thrown during cron function")

    return impl
