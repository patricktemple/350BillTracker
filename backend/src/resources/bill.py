from uuid import uuid4

from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    ForeignKey,
    Integer,
    Text,
    TypeDecorator,
    sql,
)
from sqlalchemy.dialects.postgresql import ARRAY
from sqlalchemy.orm import relationship
from ..auth import auth_required, create_jwt


from ..council_api import lookup_bill, lookup_bills
from ..utils import now
from ..models import db, TIMESTAMP, UUID
from ..views import CamelCaseSchema
from marshmallow import fields

import logging
import re
import secrets
from datetime import date, timedelta

import flask
from flask import jsonify, render_template, request
from marshmallow import fields
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import joinedload
from werkzeug import exceptions

from ..app import app


# Model -----------------------------------------------------------------------

DEFAULT_TWITTER_SEARCH_TERMS = [
    "solar",
    "climate",
    "wind power",
    "renewable",
    "fossil fuel",
]

class Bill(db.Model):
    __tablename__ = "bills"

    # These are all auto-populated by the API:
    id = Column(Integer, primary_key=True)
    file = Column(Text, nullable=False)  # e.g. Int 2317-2021
    name = Column(Text, nullable=False)
    title = Column(Text, nullable=False)
    status = Column(Text)
    body = Column(Text)

    intro_date = Column(TIMESTAMP, nullable=False)

    @property
    def tracked(self):
        return True

    # Data we track
    notes = Column(Text, nullable=False, server_default="")
    nickname = Column(Text, nullable=False, server_default="")

    twitter_search_terms = Column(
        ARRAY(Text), nullable=False, default=DEFAULT_TWITTER_SEARCH_TERMS
    )

    sponsorships = relationship(
        "BillSponsorship", back_populates="bill", cascade="all, delete"
    )
    attachments = relationship(
        "BillAttachment", back_populates="bill", cascade="all, delete"
    )
    power_hours = relationship(
        "PowerHour", back_populates="bill", cascade="all, delete"
    )

    @property
    def display_name(self):
        return self.nickname if self.nickname else self.name


class BillSchema(CamelCaseSchema):
    # Data pulled from the API
    id = fields.Integer(dump_only=True)
    name = fields.String(dump_only=True)
    title = fields.String(dump_only=True)
    status = fields.String(dump_only=True)
    body = fields.String(dump_only=True)
    file = fields.String(dump_only=True)

    # Data that we track
    tracked = fields.Boolean(dump_only=True)
    notes = fields.String(required=True)
    nickname = fields.String(required=True)
    twitter_search_terms = fields.List(fields.String(), required=True)


# Views ----------------------------------------------------------------------
from ..council_sync import update_bill_sponsorships


@app.route("/api/saved-bills", methods=["GET"])
@auth_required
def bills():
    bills = Bill.query.order_by(Bill.name).all()
    return BillSchema(many=True).jsonify(bills)


@app.route("/api/saved-bills", methods=["POST"])
@auth_required
def save_bill():
    bill_id = request.json["id"]
    if Bill.query.get(bill_id):
        # There's a race condition of checking this and then inserting,
        # but in that rare case it will hit the DB unique constraint instead.
        raise exceptions.Conflict()

    bill_data = lookup_bill(bill_id)
    logging.info(f"Saving bill {bill_id}, council API returned {bill_data}")

    bill = Bill(**bill_data)
    db.session.add(bill)

    update_bill_sponsorships(bill_id)

    db.session.commit()

    return jsonify({})


@app.route("/api/saved-bills/<int:bill_id>", methods=["PUT"])
@auth_required
def update_bill(bill_id):
    data = BillSchema().load(request.json)

    bill = Bill.query.get(bill_id)
    bill.notes = data["notes"]
    bill.nickname = data["nickname"]

    bill.twitter_search_terms = data["twitter_search_terms"]

    db.session.commit()

    return jsonify({})


@app.route("/api/saved-bills/<int:bill_id>", methods=["DELETE"])
@auth_required
def delete_bill(bill_id):
    bill = Bill.query.get(bill_id)
    db.session.delete(bill)
    db.session.commit()

    return jsonify({})


@app.route("/api/search-bills", methods=["GET"])
@auth_required
def search_bills():
    file = request.args.get("file")

    external_bills = lookup_bills(file)

    # Check whether or not we're already tracking this bill
    external_bills_ids = [b["id"] for b in external_bills]
    tracked_bills = Bill.query.filter(Bill.id.in_(external_bills_ids)).all()
    tracked_bill_ids = set([t.id for t in tracked_bills])

    for bill in external_bills:
        bill["tracked"] = bill["id"] in tracked_bill_ids

    return BillSchema(many=True).jsonify(external_bills)