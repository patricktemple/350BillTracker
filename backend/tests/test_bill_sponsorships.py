from uuid import uuid4

from src.bill.models import Bill, CityBill
from src.models import db
from src.person.models import CouncilMember, Person, Senator, AssemblyMember
from src.sponsorship.models import CitySponsorship, SenateSponsorship, AssemblySponsorship
from src.utils import now

from .utils import get_response_data


def test_get_city_bill_sponsorships(client, city_bill):
    sponsor = Person(
        id=uuid4(), name="Sponsor", type=Person.PersonType.COUNCIL_MEMBER
    )
    sponsor.council_member = CouncilMember(city_council_person_id=1)
    sponsorship = CitySponsorship(
        bill_id=city_bill.id, council_member_id=sponsor.id, sponsor_sequence=0
    )
    db.session.add(sponsor)
    db.session.add(sponsorship)

    non_sponsor = Person(
        name="Non-sponsor", type=Person.PersonType.COUNCIL_MEMBER
    )
    non_sponsor.council_member = CouncilMember(city_council_person_id=2)
    db.session.add(non_sponsor)
    db.session.commit()

    response = client.get(f"/api/city-bills/{city_bill.id}/sponsorships")

    assert response.status_code == 200
    response_data = get_response_data(response)

    assert len(response_data) == 2
    assert response_data[0]["billId"] == str(city_bill.id)
    assert response_data[0]["person"]["name"] == "Sponsor"
    assert response_data[0]["isSponsor"] == True

    assert response_data[1]["billId"] == str(city_bill.id)
    assert response_data[1]["person"]["name"] == "Non-sponsor"
    assert response_data[1]["isSponsor"] == False


def test_get_state_bill_sponsorships(client, state_bill, city_bill, snapshot, get_uuid):
    # Ugh I should really have other decoy data in there as well to make sure it's not returned...
    senate_non_sponsor = Person(id=get_uuid(),
        name="Senate non sponsor", type=Person.PersonType.SENATOR
    )
    senate_non_sponsor.senator = Senator(state_member_id=5)
    db.session.add(senate_non_sponsor)

    senate_sponsor = Person(id=get_uuid(),
        name="Senate sponsor", type=Person.PersonType.SENATOR
    )
    senate_sponsor.senator = Senator(state_member_id=2)
    db.session.add(senate_sponsor)

    assembly_non_sponsor = Person(id=get_uuid(),
        name="Assembly non sponsor", type=Person.PersonType.ASSEMBLY_MEMBER
    )
    assembly_non_sponsor.assembly_member = AssemblyMember(state_member_id=50)
    db.session.add(assembly_non_sponsor)

    assembly_sponsor = Person(id=get_uuid(),
        name="Assembly sponsor", type=Person.PersonType.ASSEMBLY_MEMBER
    )
    assembly_sponsor.assembly_member = AssemblyMember(state_member_id=1)
    db.session.add(assembly_sponsor)

    senate_sponsorship = SenateSponsorship(senate_bill_id=state_bill.id, senator_id=senate_sponsor.id)
    db.session.add(senate_sponsorship)

    assembly_sponsorship = AssemblySponsorship(assembly_bill_id=state_bill.id, assembly_member_id=assembly_sponsor.id)
    db.session.add(assembly_sponsorship)

    db.session.commit()

    response = client.get(f"/api/state-bills/{state_bill.id}/sponsorships")

    assert response.status_code == 200
    response_data = get_response_data(response)

    assert response_data == snapshot