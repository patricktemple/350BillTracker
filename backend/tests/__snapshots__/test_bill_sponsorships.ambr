# name: test_get_state_bill_sponsorships
  <class 'dict'> {
    'assemblyNonSponsors': <class 'list'> [
      <class 'dict'> {
        'assemblyMember': <class 'dict'> {
        },
        'councilMember': None,
        'email': None,
        'id': '00000000-0000-0000-0000-000000000004',
        'name': 'Assembly non sponsor',
        'notes': None,
        'party': None,
        'phone': None,
        'senator': None,
        'title': None,
        'twitter': None,
        'type': 'ASSEMBLY_MEMBER',
      },
    ],
    'assemblySponsors': <class 'list'> [
      <class 'dict'> {
        'assemblyMember': <class 'dict'> {
        },
        'councilMember': None,
        'email': None,
        'id': '00000000-0000-0000-0000-000000000005',
        'name': 'Assembly sponsor',
        'notes': None,
        'party': None,
        'phone': None,
        'senator': None,
        'title': None,
        'twitter': None,
        'type': 'ASSEMBLY_MEMBER',
      },
    ],
    'billId': '00000000-0000-0000-0000-000000000001',
    'senateNonSponsors': <class 'list'> [
      <class 'dict'> {
        'assemblyMember': None,
        'councilMember': None,
        'email': None,
        'id': '00000000-0000-0000-0000-000000000002',
        'name': 'Senate non sponsor',
        'notes': None,
        'party': None,
        'phone': None,
        'senator': <class 'dict'> {
        },
        'title': None,
        'twitter': None,
        'type': 'SENATOR',
      },
    ],
    'senateSponsors': <class 'list'> [
      <class 'dict'> {
        'assemblyMember': None,
        'councilMember': None,
        'email': None,
        'id': '00000000-0000-0000-0000-000000000003',
        'name': 'Senate sponsor',
        'notes': None,
        'party': None,
        'phone': None,
        'senator': <class 'dict'> {
        },
        'title': None,
        'twitter': None,
        'type': 'SENATOR',
      },
    ],
  }
---
